xray/sql_context.go

```diff
@@ -0,0 +1,725 @@
+	panic("not supported")
+}
+
+func (conn *driverConn) BeginTx(ctx context.Context, opts driver.TxOptions) (driver.Tx, error) {
+	var tx driver.Tx
+	var err error
+	if connCtx, ok := conn.Conn.(driver.ConnBeginTx); ok {
+		tx, err = connCtx.BeginTx(ctx, opts)
+	} else {
+		if opts.Isolation != driver.IsolationLevel(sql.LevelDefault) {
+			return nil, errors.New("xray: driver does not support non-default isolation level")
+		}
+		if opts.ReadOnly {
+			return nil, errors.New("xray: driver does not support read-only transactions")
+		}
+		tx, err = conn.Conn.Begin()
+		if err == nil {
+			select {
+			default:
+			case <-ctx.Done():
+				tx.Rollback()
+				return nil, ctx.Err()
+			}
+		}
+	}
+	if err != nil {
+		return nil, err
+	}
+	return &driverTx{Tx: tx}, nil
+}
+
+func (conn *driverConn) Exec(query string, args []driver.Value) (driver.Result, error) {
+	panic("not supported")
+}
+
+func (conn *driverConn) ExecContext(ctx context.Context, query string, args []driver.NamedValue) (driver.Result, error) {
+	execer, ok := conn.Conn.(driver.Execer)
+	if !ok {
+		return nil, driver.ErrSkip
+	}
+
+	var err error
+	var result driver.Result
+	if execerCtx, ok := conn.Conn.(driver.ExecerContext); ok {
+		err = Capture(ctx, conn.attr.dbname, func(ctx context.Context) error {
+			conn.attr.populate(ctx, query)
+			var err error
+			result, err = execerCtx.ExecContext(ctx, query, args)
+			return err
+		})
+	} else {
+		select {
+		default:
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		}
+		dargs, err0 := namedValuesToValues(args)
+		if err0 != nil {
+			return nil, err0
+		}
+		err = Capture(ctx, conn.attr.dbname, func(ctx context.Context) error {
+			conn.attr.populate(ctx, query)
+			var err error
+			result, err = execer.Exec(query, dargs)
+			return err
+		})
+	}
+	return result, err
+}
+
+func (conn *driverConn) Query(query string, args []driver.Value) (driver.Rows, error) {
+	panic("not supported")
+}
+
+func (conn *driverConn) QueryContext(ctx context.Context, query string, args []driver.NamedValue) (driver.Rows, error) {
+	queryer, ok := conn.Conn.(driver.Queryer)
+	if !ok {
+		return nil, driver.ErrSkip
+	}
+
+	var err error
+	var rows driver.Rows
+	if queryerCtx, ok := conn.Conn.(driver.QueryerContext); ok {
+		err = Capture(ctx, conn.attr.dbname, func(ctx context.Context) error {
+			conn.attr.populate(ctx, query)
+			var err error
+			rows, err = queryerCtx.QueryContext(ctx, query, args)
+			return err
+		})
+	} else {
+		select {
+		default:
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		}
+		dargs, err0 := namedValuesToValues(args)
+		if err0 != nil {
+			return nil, err0
+		}
+		err = Capture(ctx, conn.attr.dbname, func(ctx context.Context) error {
+			conn.attr.populate(ctx, query)
+			var err error
+			rows, err = queryer.Query(query, dargs)
+			return err
+		})
+	}
+	return rows, err
+}
+
+func (conn *driverConn) Close() error {
+	return conn.Conn.Close()
+}
+
+// copied from sql/driver/convert.go
```

Comment:
sql/driver/convert.go – is from this repo or else where? If elsewhere, do we have to care about their license?

Reply:
it is from https://github.com/golang/go/blob/e6ebbe0d20fe877b111cf4ccf8349cba129d6d3a/src/database/sql/convert.go#L93-L99